pkg/controller/hvpa/hvpa_controller.go

```diff
@@ -0,0 +1,651 @@
+					// If there is no such value, we can not scale out anyway, so we will end up with vpaWeight = 1
+					lookupNextFraction = true
+					continue
+				}
+			}
+			break
+		}
+	}
+
+	// Memory for newDeploy is assigned in the function getWeightedRequests
+	newDeploy, resourceChanged, err := getWeightedRequests(vpaStatus, hvpa, vpaWeight, deployment)
+	if err != nil {
+		log.Error(err, "Error in getting weight based requests in new deployment")
+	}
+
+	weightedReplicas, err := getWeightedReplicas(hpaStatus, hvpa, deployment, 1-vpaWeight)
+	if err != nil {
+		log.Error(err, "Error in getting weight based replicas")
+	}
+
+	if weightedReplicas == 0 {
+		weightedReplicas = currentReplicas
+	}
+
+	if currentReplicas != weightedReplicas {
+		if newDeploy == nil {
+			newDeploy = deployment.DeepCopy()
+		}
+		newDeploy.Spec.Replicas = &weightedReplicas
+
+		log.Info("HPA", "Scale horizontally from", currentReplicas, "to", weightedReplicas)
+	}
+	if newDeploy == nil {
+		log.Info("Scaling not required")
+		return 0, false, vpaWeight, nil
+	}
+	log.Info("Scaling required")
+	return weightedReplicas - currentReplicas, resourceChanged, vpaWeight, r.Update(context.TODO(), newDeploy)
+}
+
+func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment, hpaWeight autoscalingv1alpha1.VpaWeight) (int32, error) {
+	anno := hvpa.GetAnnotations()
+	if val, ok := anno["hpa-controller"]; ok {
+		if val == "kcm" {
+			// HPA is controlled by kube-controller-manager
+			log.Info("HVPA controller is not controlling HPA")
+			return 0, nil
+		}
+	}
+
+	log.Info("Calculating weighted replicas")
+	if hpaWeight == 0 || hpaStatus == nil || hpaStatus.DesiredReplicas == 0 {
+		log.Info("Nothing to do")
+		return 0, nil
+	}
+
+	lastScaleTime := hvpa.Status.HvpaStatus.LastScaleTime
+	if lastScaleTime == nil {
+		lastScaleTime = &metav1.Time{}
+	}
+	lastScaleTimeDuration := time.Now().Sub(lastScaleTime.Time)
+	scaleUpDelay, _ := time.ParseDuration(hvpa.Spec.ScaleUpDelay)
+	scaleDownDelay, _ := time.ParseDuration(hvpa.Spec.ScaleDownDelay)
+
+	var err error
+	minReplicas := *hvpa.Spec.HpaTemplate.MinReplicas
+	maxReplicas := hvpa.Spec.HpaTemplate.MaxReplicas
+	currentReplicas := *deployment.Spec.Replicas
+	desiredReplicas := hpaStatus.DesiredReplicas
+
+	weightedReplicas := int32(math.Ceil(float64(currentReplicas) + float64(desiredReplicas-currentReplicas)*float64(hpaWeight)))
+	if weightedReplicas < minReplicas {
+		weightedReplicas = minReplicas
+	}
+	if weightedReplicas > maxReplicas {
+		weightedReplicas = maxReplicas
+	}
+
+	if (weightedReplicas > currentReplicas && lastScaleTimeDuration > scaleUpDelay) ||
+		(weightedReplicas < currentReplicas && lastScaleTimeDuration > scaleDownDelay) {
+		log.Info("HPA", "weighted replicas", weightedReplicas)
+		return weightedReplicas, err
+	}
+
+	log.Info("HPA", "no scaling done. Current replicas", currentReplicas)
```

Comment:
Might be good to log (or even a condition in status) why scaling was not done. Because recommendations did not change or because the stabilisation window was effective?